core/src/test/scala/unit/kafka/controller/ControllerFailoverTest.scala

```diff
@@ -54,119 +53,47 @@ class ControllerFailoverTest extends KafkaServerTestHarness with Logging {
-          log.info("Queue state %d %d".format(channelManager.queueCapacity(0), channelManager.queueSize(0)))
-          controller.kafkaController.sendUpdateMetadataRequest(Seq(0), Set(topicPartition))
-          log.info("Queue state %d %d".format(channelManager.queueCapacity(0), channelManager.queueSize(0)))
-        } catch {
-          case _: Exception => log.info("Thread interrupted")
-        }
+    TestUtils.waitUntilTrue(() =>
+      initialController.partitionStateMachine.partitionsInState(OnlinePartition).contains(topicPartition),
+      s"Partition $topicPartition did not transition to online state")
+
+    // Wait until we have verified that we have resigned
+    val latch = new CountDownLatch(1)
+    @volatile var expectedExceptionThrown = false
+    @volatile var unexpectedExceptionThrown: Option[Throwable] = None
+    val illegalStateEvent = ControllerTestUtils.createMockControllerEvent(ControllerState.BrokerChange, { () =>
+      try initialController.handleIllegalState(new IllegalStateException("Thrown for test purposes"))
+      catch {
+        case _: IllegalStateException => expectedExceptionThrown = true
+        case t: Throwable => unexpectedExceptionThrown = Some(t)
       }
+      latch.await()
     })
-    thread.setName("mythread")
-    thread.start()
-    while (thread.getState() != Thread.State.WAITING) {
-      Thread.sleep(100)
-    }
-    // Assume that the thread is WAITING because it is
-    // blocked on the queue, so interrupt and move forward
-    thread.interrupt()
-    thread.join()
-    channelManager.resumeSendThread(0)
-    // Wait and find current controller
-    var found = false
-    var counter = 0
-    while (!found && counter < 10) {
-      for (server <- this.servers) {
-        val previousEpoch = epochMap get server.config.brokerId match {
-          case Some(epoch) =>
-            epoch
-          case None =>
-            val msg = String.format("Missing element in epoch map %s", epochMap.mkString(", "))
-            throw new IllegalStateException(msg)
-        }
+    initialController.eventManager.put(illegalStateEvent)
+    // Check that we have shutdown the scheduler (via onControllerResigned)
+    TestUtils.waitUntilTrue(() => !initialController.kafkaScheduler.isStarted, "Scheduler was not shutdown")
+    TestUtils.waitUntilTrue(() => !initialController.isActive, "Controller did not become inactive")
+    latch.countDown()
+    TestUtils.waitUntilTrue(() => expectedExceptionThrown, "IllegalStateException was not thrown")
+    assertEquals("Unexpected exception thrown", None, unexpectedExceptionThrown)
 
-        if (server.kafkaController.isActive
-            && previousEpoch < server.kafkaController.epoch) {
-          controller = server
-          found = true
+    TestUtils.waitUntilTrue(() => {
+      servers.exists { server =>
+        val previousEpoch = epochMap.get(server.config.brokerId).getOrElse {
```

Comment:
It seems that we don't really need epochMap. previousEpoch could be obtained from the previous controller. If a broker is not a controller, the controller epoch is not valid.